runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -252,6 +252,249 @@ doubleMaxMinHelper(TR::Node *node, TR::CodeGenerator *cg, bool isMaxOp)
    return node->getRegister();
    }
 
+  /** \brief
+   *     Attempts to use vector registers to perform SIMD conversion of characters from lowercase to uppercase
+   *
+   *  \detail
+   *     Uses vector registers to convert 16 bytes at a time
+   *
+   *  \param node
+   *     The node representing the HW optimized toUpper and toLower recognized calls
+   *
+   *  \param cg
+   *     The code generator used to generate the instructions
+   *
+   *  \param isToUpper
+   *     Boolean representing case conversion, either to upper or to lower
+   *
+   *  \param isCompressedString
+   *     Boolean representing the string's compression
+   *
+   *  \return
+   *     0 if the conversion is unsuccessful, a non zero value otherwise
+   */
+TR::Register * caseConversionHelper(TR::Node* node, TR::CodeGenerator* cg, bool isToUpper, bool isCompressedString)
+   {
+   TR::Register* sourceRegister = cg->evaluate(node->getChild(0));
+   TR::Register* destRegister = cg->evaluate(node->getChild(1));
+   TR::Register* lengthRegister = cg->gprClobberEvaluate(node->getChild(2));
+
+   TR::Register* addressOffset = cg->allocateRegister();
+   TR::Register* loadLength = cg->allocateRegister();
+   // Loopcounter register for number of 16 byte conversions, when it is used, the length is not needed anymore
+   TR::Register* loopCounter = lengthRegister;
+
+   TR::Register* charBufferVector = cg->allocateRegister(TR_VRF);
+   TR::Register* selectionVector = cg->allocateRegister(TR_VRF);
+   TR::Register* modifiedCaseVector = cg->allocateRegister(TR_VRF);
+   TR::Register* charOffsetVector = cg->allocateRegister(TR_VRF);
+   TR::Register* alphaRangeVector = cg->allocateRegister(TR_VRF);
+   TR::Register* alphaCondVector = cg->allocateRegister(TR_VRF);
+   TR::Register* invalidRangeVector = cg->allocateRegister(TR_VRF);
+   TR::Register* invalidCondVector = cg->allocateRegister(TR_VRF);
+
+   TR::LabelSymbol* fullVectorConversion = TR::LabelSymbol::create(cg->trHeapMemory(), cg);
+   TR::LabelSymbol* doneLabel = TR::LabelSymbol::create(cg->trHeapMemory(), cg);
+   TR::LabelSymbol* handleInvalidChars = TR::LabelSymbol::create(cg->trHeapMemory(), cg);
+   TR::LabelSymbol* loop = TR::LabelSymbol::create(cg->trHeapMemory(), cg);
+
+   TR::Instruction* instruction;
+
+   const int elementSizeMask = (isCompressedString) ? 0x0 : 0x1;    // byte or halfword mask
+   const int sizeOfVector = 16;
+   const bool is64 = TR::Compiler->target.is64Bit();
+   uintptrj_t headerSize = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+
+   TR::RegisterDependencyConditions * regDeps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 14, cg);
```

Comment:
I only count 13 dependencies below. Can we save some space?

Reply:
5143440